sticker-utils/src/subcommands/distill.rs

```diff
@@ -0,0 +1,721 @@
+        let weight_decay = matches
+            .value_of(WEIGHT_DECAY)
+            .unwrap()
+            .parse()
+            .or_exit("Cannot parse weight decay", 1);
+
+        // If steps is present, it overrides epochs.
+        let train_duration = if let Some(steps) = matches.value_of(STEPS) {
+            let steps = steps
+                .parse()
+                .or_exit("Cannot parse the number of training steps", 1);
+            TrainDuration::Steps(steps)
+        } else {
+            let epochs = matches
+                .value_of(EPOCHS)
+                .unwrap()
+                .parse()
+                .or_exit("Cannot parse number of training epochs", 1);
+            TrainDuration::Epochs(epochs)
+        };
+
+        DistillApp {
+            batch_size,
+            device,
+            eval_steps,
+            hard_loss,
+            max_len,
+            lr_schedules: RefCell::new(Self::create_lr_schedules(
+                initial_lr_classifier,
+                initial_lr_encoder,
+                lr_decay_rate,
+                lr_decay_steps,
+                warmup_steps,
+            )),
+            student_config,
+            teacher_config,
+            train_data,
+            train_duration,
+            validation_data,
+            weight_decay,
+        }
+    }
+
+    fn run(&self) {
+        let student_config = load_config(&self.student_config);
+        let teacher = Model::load(&self.teacher_config, self.device, true);
+
+        let train_file = File::open(&self.train_data).or_exit("Cannot open train data file", 1);
+        let mut validation_file =
+            File::open(&self.validation_data).or_exit("Cannot open validation data file", 1);
+
+        let student = self.fresh_student(&student_config, &teacher);
+
+        let mut optimizer = AdamW::new(&student.vs);
+
+        self.distill_model(
+            &mut optimizer,
+            &teacher,
+            &student,
+            &train_file,
+            &mut validation_file,
+        )
+        .or_exit("Model distillation failed", 1);
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+enum TrainDuration {
+    Epochs(usize),
+    Steps(usize),
+}
+
+impl TrainDuration {
+    fn to_steps(&self, train_file: &File, batch_size: usize) -> Fallible<usize> {
+        use TrainDuration::*;
+
+        match *self {
+            Epochs(epochs) => {
+                eprintln!("Counting number of steps in an epoch...");
+                let read_progress =
+                    ReadProgress::new(train_file.try_clone()?).or_exit("Cannot open train file", 1);
+
+                let progress_bar = read_progress.progress_bar().clone();
+                progress_bar
+                    .set_style(ProgressStyle::default_bar().template(
+                        "[Time: {elapsed_precise}, ETA: {eta_precise}] {bar} {percent}%",
+                    ));
+
+                let n_sentences = count_conllx_sentences(BufReader::new(read_progress))?;
```

Comment:
Is this full pass over the train data done to obtain a progress bar?

Reply:
We could get a progress bar using just the `Read + Seek` (as is done in finetune), though since we have it, it is also used for that.

The sentence count is done to count how many steps an epoch is, so that the user can specify `--epochs N`, since the main loop is step/batch-driven.

Reply:
Ah, wouldn't epoch-signalling also be possible by returning a None batch at the end of the file? I can also live with what it is now, the one-off read-through is probably dwarfed by the training time anyways. I guess it's a minor annoyance when working with really large files.

Reply:
It's really cheap. Usually < 1 minute, since it does not actually parse the CoNLL-X, but just counts blank lines.